pkg/syncer/endpoints/endpoint_downstream_controller.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpoints
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	ddsif "github.com/kcp-dev/kcp/pkg/informer"
+	"github.com/kcp-dev/kcp/pkg/logging"
+)
+
+const (
+	ControllerName = "syncer-endpoint-controller"
+)
+
+// NewEndpointController returns new controller which would annotate Endpoints related to synced Services, so that those Endpoints
+// would be upsynced by the UpSyncer to the upstream KCP workspace.
+// This would be useful to enable components such as a KNative controller (running against the KCP workspace) to see the Endpoint,
+// and confirm that the related Service is effective.
+func NewEndpointController(
+	downstreamClient dynamic.Interface,
+	ddsifForDownstream *ddsif.GenericDiscoveringDynamicSharedInformerFactory[cache.SharedIndexInformer, cache.GenericLister, informers.GenericInformer],
+) (*controller, error) {
+	endpointsGVR := corev1.SchemeGroupVersion.WithResource("endpoints")
+
+	c := &controller{
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName),
+	}
+
+	endpointsInformer, ok := ddsifForDownstream.Informer(endpointsGVR)
+	if !ok {
+		return nil, errors.New("endpoints informer should be available")
+	}
+
+	endpointsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
remind me what the logic is that this handler gets initial add events?

Reply:
When this handler is added, the informer is already started. However, the AddEventHandler method supports this, and explicitly notifies initial add events as shown by the following code snippet:

https://github.com/kubernetes/client-go/blob/v0.24.9/tools/cache/shared_informer.go#L547-L563